### ISSUE TITLE
ConcurrentModificationException in MetaDataStore.keysDataToJson for Crashlytics

### ISSUE THREAD

**Issue by amal** [firebase/firebase-android-sdk#2840]
### Environment

  * Android Studio version: Arctic Fox | 2020.3.1 Beta 5
  * Firebase Component: **Crashlytics**
  * Component version: **18.1.0**

### Problem

Normal usage of Crashlytics **v18.1.0** leads to error in logcat:

```stacktrace
E/FirebaseCrashlytics: Error serializing key/value metadata.
    java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1475)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1473)
        at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1708)
        at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1701)
        at org.json.JSONObject.<init>(JSONObject.java:138)
        at com.google.firebase.crashlytics.internal.common.MetaDataStore.keysDataToJson(MetaDataStore.java:157)
        at com.google.firebase.crashlytics.internal.common.MetaDataStore.writeKeyData(MetaDataStore.java:91)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController$8.call(CrashlyticsController.java:484)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController$8.call(CrashlyticsController.java:480)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsBackgroundWorker$3.then(CrashlyticsBackgroundWorker.java:105)
        at com.google.android.gms.tasks.zzc.run(com.google.android.gms:play-services-tasks@@17.2.1:3)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at com.google.firebase.crashlytics.internal.common.ExecutorUtils$1$1.onRun(ExecutorUtils.java:64)
        at com.google.firebase.crashlytics.internal.common.BackgroundPriorityRunnable.run(BackgroundPriorityRunnable.java:27)
        at java.lang.Thread.run(Thread.java:920)
```

#### Steps to reproduce:

Issue happens on application start.
`FirebaseCrashlytics#setUserId` and `FirebaseCrashlytics#log` methods used from application code.

**Comment by aguatno** [comment]
Hi @amal thanks for reporting. I checked our system and found no similar issue reported. Are you able to reproduce this with the sample [quickstart app](https://github.com/firebase/quickstart-android) and the most recent version of Firebase? A sample project where the issue can be reproduced will significantly help the investigation. Thanks

**Comment by amal** [comment]
@aguatno, thank you. I'll try to reproduce issue in a quickstart app. But it will take some time and can be tricky as the problem occurs sporadically in my application.

**Comment by sjr** [comment]
In case it is helpful: we are seeing the same internal crash with firebase 17.4.1

**Comment by ivangonzalez88** [comment]
It looks like the reason for the issue is that [KeysMap](https://github.com/firebase/firebase-android-sdk/blob/1bb91c157661220cc67df1d0918d76785ae8872e/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/KeysMap.java#L26)  is not thread safe.

The below java code will generate ConcurrentModificationException
```java
class Main {

    public static void main(String[] args) {
        KeysMap keysMap = new KeysMap(1000, 10000);
        int i = 0;
        while (i++ < 1000) {
            keysMap.setKey(i + "", "");
            new Thread(new Runnable() {
                @Override
                public void run() {
                    for (Map.Entry<String, String> entry : keysMap.getKeys().entrySet()) {}
                }
            }).start();
            ;
        }
    }
}
```

**Comment by mrichards** [comment]
Hi folks, I'm able to reproduce the bug internally and we'll have a fix in an upcoming release. Note this does NOT cause the app to crash, but it could result in inconsistent / unexpected data when using Crashlytics' custom key/data feature. (If you are seeing app crashes due to this, please report them here!)


While it appears that there's a long-time bug that could result in a `ConcurrentModificationException` under certain circumstance (both for Unity and non-Unity apps), there was a change in the Firebase Unity 7.2.0 release that makes it considerably more likely to occur for Unity apps, especially near app launch.

(@ivangonzalez88 your repro example won't compile against the latest release of the SDK because of recent, soon-to-be-shipped source changes. But the general principal to reproduce the bug is the same.)

**Comment by mrichards** [comment]
Hi folks, this issue is fixed as of v18.2.1!
